src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -56,8 +56,417 @@ private static unsafe void ThrowExceptionEncryptDecryptFail(string fullPath)
             throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
         }
 
+        private static SafeFileHandle? TryOpenTargetNoAlternativeDataStream(
+            string lpFileName,
+            int dwDesiredAccess,
+            FileShare dwShareMode,
+            FileMode dwCreationDisposition,
+            int dwFlagsAndAttributes,
+            ref Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA fileInformation,
+            out string openedFileName)
+        {
+            // Default value for openedFileName.
+            openedFileName = lpFileName;
+
+            // Check the file isn't an alternative data stream
+            if (Path.GetFileName(lpFileName.AsSpan()).Contains(':'))
+            {
+                return null;
+            }
+
+            // Try to open it in a way where we can detect symlinks.
+            SafeFileHandle result = Interop.Kernel32.CreateFile(
```

Comment:
Consider we have a file, we check that it's not a symlink by reading the attributes, and then we open and return it. BUT between reading the attributes and opening the handle, it gets deleted and replaced with a symlink to an ADS. Now we get a BSOD. That's why I've ordered it to open a handle first (which should be locked for correct usage of this API), then read the attributes.
Does that make sense? Should probably add a comment explaining this?

Reply:
Note: I am assuming that an open file handle would stop this, would be good to confirm.

Reply:
I think we can actually get rid of it if we later call `GetFileInformationByHandleEx` with `FileNameInfo` to extract the filename (when we need it, for the call to `GetVolumePathName`), that way we can skip specifying `FILE_FLAG_OPEN_REPARSE_POINT` and just extract the name later. It will also reduce the number of syscalls on the golden path to failure to 2 (which is almost certainly the minimum amount).

We will have to check ADS at some later point, but that's fine since they're an edge case imo, and it's only `DUPLICATE_EXTENTS_DATA` which causes the issue.